File: problems/permutations/backtracking.cs
public class Solution {
    public IList<IList<int>> Permute(int[] nums) {
        var results = new List<IList<int>>();
        var subset = new List<int>();

        Backtrack(nums, subset, results);

        return results;
    }

    private void Backtrack(int[] nums, List<int> subset, List<IList<int>> results)
    {
        if (subset.Count == nums.Length)
        {
            results.Add(new List<int>(subset));
            return;
        }

        for (var i = 0; i < nums.Length; i++)
        {
            var num = nums[i];
            if (subset.Contains(num)) continue;
            
            subset.Add(num);

            Backtrack(nums, subset, results);

            subset.RemoveAt(subset.Count - 1);
        }
    }
}

File: problems/permutations/backtracking.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        results = []
        subset = []
        visited = set()

        self.backtrack(0, nums, visited, subset, results)

        return results

    def backtrack(self, start: int, nums: List[int], visited: set, subset: List[int], results: List[List[int]]):
        if len(subset) == len(nums):
            results.append([x for x in subset])
            return

        for i in range(0, len(nums)):
            num = nums[i]
            if num in visited:
                continue
            
            subset.append(num)
            visited.add(num)

            self.backtrack(i, nums, visited, subset, results)

            subset.pop()
            visited.remove(num)